solr/contrib/ltr/src/java/org/apache/solr/ltr/ranking/ModelQuery.java

```diff
@@ -259,43 +262,79 @@ public String toString(String field) {
     return field;
   }
 
+  public class FeatureInfo {
+      String name;
+      float value;
+      boolean used;
+      
+      FeatureInfo(String n, float v, boolean u){
+        name = n; value = v; used = u;
+      }
+       
+      public boolean getUsed(){
+        return used;
+      }
+      
+      public void setUsed(boolean u){
+        used = u;
+      }
+      
+      public void setScoreUsed(float score, boolean used){
+        this.value = score;
+        this.used = used;
+      }
+      
+      public String getName(){
+        return name;
+      }
+      
+      public float getValue(){
+        return value;
+      }
+  }
+  
   public class ModelWeight extends Weight {
 
     IndexSearcher searcher;
 
     // List of the model's features used for scoring. This is a subset of the
     // features used for logging.
-    FeatureWeight[] modelFeatures;
+    FeatureWeight[] modelFeatureWeights;
     float[] modelFeatureValuesNormalized;
+    FeatureWeight[] extractedFeatureWeights;
+    Normalizer[] modelFeatureNorms;
 
-    // List of all the feature values, used for both scoring and logging
-    FeatureWeight[] allFeatureWeights;
-    float[] allFeatureValues;
-    String[] allFeatureNames;
-    boolean[] allFeaturesUsed;
-    HashMap<Integer, Normalizer> modelFeatureNorms;
+    // List of all the feature names, values, used for both scoring and logging
+    LinkedHashMap<Integer, FeatureInfo> featuresInfo;
```

Comment:
What is the advantage of using a hashmap here instead of an array of objects?
    A set of arrays was used earlier and the elements were accessed using the featureId. With the updated logic, the number of elements in the array can be fewer than the  total number of features. When [features] are not requested, only the model features are extracted. In this case, the indexing by featureId, fails. For this reason, we need a map which holds just the features that were triggered by the documents in the result set. The 'used' flag in the FeatureInfo object, indicates that the feature was triggered during the scoring.